File: leku/src/main/java/com/schibstedspain/leku/geocoder/GeocoderRepository.java
package com.schibstedspain.leku.geocoder;

import android.location.Address;
import com.google.android.gms.maps.model.LatLng;
import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import java.util.List;

public class GeocoderRepository {
  private static final int RETRY_COUNT = 3;

  private final GeocoderInteractorDataSource androidGeocoder;
  private final GeocoderInteractorDataSource googleGeocoder;

  public GeocoderRepository(GeocoderInteractorDataSource androidGeocoder, GeocoderInteractorDataSource googleGeocoder) {
    this.androidGeocoder = androidGeocoder;
    this.googleGeocoder = googleGeocoder;
  }

  public Observable<List<Address>> getFromLocationName(String query) {
    return androidGeocoder.getFromLocationName(query)
        .subscribeOn(Schedulers.newThread())
        .retry(RETRY_COUNT)
        .onErrorResumeNext(googleGeocoder.getFromLocationName(query))
        .onExceptionResumeNext(googleGeocoder.getFromLocationName(query));
  }

  public Observable<List<Address>> getFromLocationName(String query, LatLng lowerLeft, LatLng upperRight) {
    return androidGeocoder.getFromLocationName(query, lowerLeft, upperRight)
        .subscribeOn(Schedulers.newThread())
        .retry(RETRY_COUNT)
        .onErrorResumeNext(googleGeocoder.getFromLocationName(query, lowerLeft, upperRight))
        .onExceptionResumeNext(googleGeocoder.getFromLocationName(query, lowerLeft, upperRight));
  }

  public Observable<List<Address>> getFromLocation(LatLng latLng) {
    return androidGeocoder.getFromLocation(latLng.latitude, latLng.longitude)
        .subscribeOn(Schedulers.newThread())
        .retry(RETRY_COUNT)
        .onErrorResumeNext(googleGeocoder.getFromLocation(latLng.latitude, latLng.longitude))
        .onExceptionResumeNext(googleGeocoder.getFromLocation(latLng.latitude, latLng.longitude));
  }
}

File: leku/src/main/java/com/schibstedspain/leku/geocoder/GeocoderRepository.kt
package com.schibstedspain.leku.geocoder

import android.location.Address
import com.google.android.gms.maps.model.LatLng
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers

private const val RETRY_COUNT = 3

class GeocoderRepository(
    private val androidGeocoder: GeocoderInteractorDataSource,
    private val googleGeocoder: GeocoderInteractorDataSource
) {

    fun getFromLocationName(query: String): Observable<List<Address>> {
        return androidGeocoder.getFromLocationName(query)
                .subscribeOn(Schedulers.newThread())
                .retry(RETRY_COUNT.toLong())
                .onErrorResumeWith(googleGeocoder.getFromLocationName(query))
    }

    fun getFromLocationName(query: String, lowerLeft: LatLng, upperRight: LatLng): Observable<List<Address>> {
        return androidGeocoder.getFromLocationName(query, lowerLeft, upperRight)
                .subscribeOn(Schedulers.newThread())
                .retry(RETRY_COUNT.toLong())
                .onErrorResumeWith(googleGeocoder.getFromLocationName(query, lowerLeft, upperRight))
    }

    fun getFromLocation(latLng: LatLng): Observable<List<Address>> {
        return androidGeocoder.getFromLocation(latLng.latitude, latLng.longitude)
                .subscribeOn(Schedulers.newThread())
                .retry(RETRY_COUNT.toLong())
                .onErrorResumeWith(googleGeocoder.getFromLocation(latLng.latitude, latLng.longitude))
    }
}
